### PR TITLE
Add tag filter to index page with tests

### LINKED ISSUE #6: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [ ] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by taowen19**
Closes #6

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -72,12 +104,14 @@ def test_create__failure
...
   def test_show__success
-    image = Image.create!(url: @valid_image_url, id: 1, tag_list: 'tag1 tag2')
+    image = Image.create!(url: @valid_image_url, id: 1, tag_list: %w[tag1 tag2])
     get image_url(image)
     assert_response :success
 
-    assert_select 'h1', 'tag1 tag2'
-
+    assert_select 'div' do
+      assert_select 'a[href=?]', @default_domain + '/images?tag=tag1'
```
Comment (revatidamle) [inline test/controllers/images_controller_test.rb:112]: what is the purpose of `@default_domain` here?

File: app/views/images/index.html.erb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <h1>Image</h1>
-
+<%= content_tag(:p, "Nonexistent Images") if @images.blank? %>
```
Comment (revatidamle) [inline app/views/images/index.html.erb:2]: Good thinking to show a message when there are no images! 
Maybe a more descriptive message could be given? `There are no images to be displayed`

File: app/views/images/index.html.erb [hunk 3]
```diff
@@ -8,7 +8,7 @@
       </td>
       <td>
         <% image.tag_list.each do |tag| %>
-          <p><%= tag %></p>
+          <%= link_to tag, images_url(tag: tag) %>
```
Comment (revatidamle) [inline app/views/images/index.html.erb:11]: We should be using `images_path` that is provided by the rails route. `_path` provides a relative path, whereas `_url` gives the absolute path. `_url` is mostly used in cases where you want to redirect from external services. You can check out https://guides.rubyonrails.org/routing.html for more information

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (revatidamle) [inline test/controllers/images_controller_test.rb:112]: if you replace `images_url` with `image_path` then you might not need to set a default domain

Comment (taowen19) [reply test/controllers/images_controller_test.rb:112]: > if you replace `images_url` with `image_path` then you might not need to set a default domain

Just an update for this one. I'm using image-path now and src shows relative paths when I inspect in my browser. But when I run tests locally, <a> tags are selected and printed out with absolute paths, such as '<a href="http://www.example.com/images?tag=tag1">tag1</a>'. Perhaps dev and test has different behaviors?

Comment (taowen19) [reply test/controllers/images_controller_test.rb]: Update: Get rid of default-domain by applying regex

File: app/views/images/show.html.erb [hunk 4]
```diff
@@ -1,3 +1,8 @@
 <a href="/">Home Page</a>
-<h1><%= @image.tag_list %></h1>
+<div>
+  <% @image.tag_list.each do |tag| %>
+      <%= link_to tag, images_url(tag: tag) %>
```
Comment (revatidamle) [inline app/views/images/show.html.erb:4]: update to `images_path` here as well

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -66,18 +84,22 @@ def test_create__failure
...
     assert_response :success
-
-    assert_select 'h1', 'tag1 tag2'
-
+    assert_select 'div' do
+      assert_select 'a:first-child' do
+        assert_select "a:match('href',?)", %r{.*/images\?tag=tag1}
+      end
+      assert_select 'a:last-child' do
+        assert_select "a:match('href',?)", %r{.*/images\?tag=tag2}
```
Comment (revatidamle) [inline test/controllers/images_controller_test.rb:100]: 💯

File: app/views/images/show.html.erb (hunk 4 above)
Comment (taowen19) [reply app/views/images/show.html.erb]: Thanks! Great catch! I've double-checked and corrected all appearances in my latest commit.

**Comment by taowen19** [comment]
[Filter Images by Tag](https://trello.com/c/PyEaGzjT/2775-filter-images-by-tag)

**Review by revatidamle — APPROVED** [review]

**Pull request merged**